File: webapp/apps/regolia/src/models/entities/year/year-level-speciality.ts
import {LevelSpeciality} from "../organisation";
import {Entity} from "../entity";
import {YearSpeciality} from "./year-speciality";
import {YearLevel} from "./year-level";
import {SemesterLevelSpeciality} from "../semester";
import {YearStudent} from "./year-student";

export class YearLevelSpeciality extends Entity<number> {

  constructor(value: any = {}) {
    super();

    this.id = value.id;
    this.registrationDate = new Date(value.registrationDate);

    this.yearSpecialityId = value.yearSpecialityId;
    this.levelSpecialityId = value.levelSpecialityId;

    if(value.levelSpeciality){
      this.levelSpeciality = new LevelSpeciality(value.levelSpeciality);
    }

    this.yearLevelId = value.yearLevelId;
  }

  yearLevelId: number;
  yearLevel: YearLevel;

  yearSpecialityId: number;
  yearSpeciality: YearSpeciality;

  levelSpeciality: LevelSpeciality;
  levelSpecialityId: number;

  semesterLevelSpecialities: SemesterLevelSpeciality[];
  yearStudents: YearStudent[];

  url(path?: string): string {
    const url = `${this.yearLevel.url()}/specialities/${this.id}`;
    if(path) {
      return `${url}/path`;
    }
    return url;
  }
}

File: ServerApp/Entities/Periods/Year/YearLevelSpeciality.cs
using System.Collections.Generic;
using Everest.AspNetStartup.Binding;
using Everest.AspNetStartup.Models;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;

namespace Exam.Entities.Periods
{
    [ModelBinder(BinderType = typeof(ItemValueModelBinder))]
    public class YearLevelSpeciality: PeriodChild
    {
        public virtual LevelSpeciality LevelSpeciality { get; set; }
        public long? LevelSpecialityId { get; set; }


        [JsonIgnore]
        public virtual YearLevel YearLevel { get; set; }
        public long? YearLevelId { get; set; }

        [JsonIgnore]
        public virtual YearSpeciality YearSpeciality { get; set; }
        public long? YearSpecialityId { get; set; }

        [JsonIgnore]
        public virtual List<SemesterLevelSpeciality> SemesterLevelSpecialities { get; set; }
        
        [JsonIgnore]
        public virtual List<YearStudent> YearStudents { get; set; }
    }
}